unittests/AST/ASTImporterTest.cpp

```diff
@@ -5171,6 +5171,102 @@ TEST_P(ASTImporterOptionSpecificTestBase, ImportOfFriendTemplateWithArgExpr) {
   EXPECT_NE(ToD1->getCanonicalDecl(), ToD2->getCanonicalDecl());
 }
 
+struct ASTImporterCImplicitFunctionTest : ASTImporterOptionSpecificTestBase {
+  static FunctionDecl *GetImplicitF(FunctionDecl *F) {
+    return cast<CallExpr>(cast<CompoundStmt>(F->getBody())->body_front())
+        ->getDirectCallee();
+  };
+};
+
+TEST_P(ASTImporterCImplicitFunctionTest, ImportCImplicitFunction) {
+  auto Code = "void f() { implicit_f(); }";
+
+  TranslationUnitDecl *FromTU = getTuDecl(Code, Lang_C, "input.c");
+  auto *FromF = FirstDeclMatcher<FunctionDecl>().match(
+      FromTU, functionDecl(hasName("f")));
+  FunctionDecl *FromImplicitF = GetImplicitF(FromF);
+
+  ASSERT_TRUE(FromImplicitF->getDeclContext() == cast<DeclContext>(FromTU));
+  ASSERT_TRUE(FromImplicitF->getLexicalDeclContext() ==
+              cast<DeclContext>(FromF));
+  ASSERT_FALSE(FromTU->containsDecl(FromImplicitF));
+  ASSERT_TRUE(FromF->containsDecl(FromImplicitF));
+
+  auto *ToImplicitF = Import(FromImplicitF, Lang_C);
+  auto *ToF = Import(FromF, Lang_C);
+  ASSERT_TRUE(ToImplicitF);
+  TranslationUnitDecl *ToTU = ToImplicitF->getTranslationUnitDecl();
+  EXPECT_FALSE(ToTU->containsDecl(ToImplicitF));
+  EXPECT_TRUE(ToF->containsDecl(ToImplicitF));
+}
+
+TEST_P(ASTImporterCImplicitFunctionTest, ImportedCImplicitFunctionsLinked) {
+  auto Code =
+      R"(
+        void f1() { implicit_f(); }
+        void f2() { implicit_f(); }
+        )";
+
+  TranslationUnitDecl *FromTU = getTuDecl(Code, Lang_C, "input.c");
+  auto *FromF1 = FirstDeclMatcher<FunctionDecl>().match(
+      FromTU, functionDecl(hasName("f1")));
+  FunctionDecl *FromImplicitF1 = GetImplicitF(FromF1);
+  auto *FromF2 = FirstDeclMatcher<FunctionDecl>().match(
+      FromTU, functionDecl(hasName("f2")));
+  FunctionDecl *FromImplicitF2 = GetImplicitF(FromF2);
+
+  ASSERT_EQ(FromImplicitF1, FromImplicitF2);
+
+  auto *ToImplicitF1 = Import(FromImplicitF1, Lang_C);
+  auto *ToF1 = Import(FromF1, Lang_C);
+  auto *ToF2 = Import(FromF2, Lang_C);
+  ASSERT_TRUE(ToImplicitF1);
+  EXPECT_EQ(GetImplicitF(ToF1), ToImplicitF1);
+  EXPECT_EQ(GetImplicitF(ToF2), ToImplicitF1);
```

Comment:
Should there be a redecl chain here?
I'd expect that when we import f2 then we find the existing `implicit_f` so the newly imported `implict_f` should be linked to that as an additional decl  in the same redecl chain.

Reply:
Yes I would expect it too but the AST dump shows that there is no chain. In this code the same `FunctionDecl` is used in `f1` and `f2` (source location returned by getLocation is valid and points to the first occurrence in `f1`).
I do not know if it is problem is we create the redecl chain at import even if originally there is no chain. This would mean create a new `FunctionDecl` instead of the `MapImported` in ASTImporter. (Only to have less special cases in the code that may cause other new problems.)

Reply:
If the "from" AST does not have a redecl chain, then probably we should not have either, but then we should have an assert_xx for the no chain in the "from" ctx.